backend/danswer/connectors/teams/connector.py

```diff
@@ -0,0 +1,288 @@
+
+logger = setup_logger()
+
+datetime_format_string = "%Y-%m-%dT%H:%M:%S.%f%z"
+
+
+def get_created_datetime(chat_message: ChatMessage) -> datetime:
+    # Extract the 'createdDateTime' value from the 'properties' dictionary and convert it to a datetime object
+    return datetime.strptime(
+        chat_message.properties["createdDateTime"], datetime_format_string
+    )
+
+
+class TeamsConnector(LoadConnector, PollConnector):
+    def __init__(
+        self,
+        batch_size: int = INDEX_BATCH_SIZE,
+        teams: list[str] = [],
+    ) -> None:
+        self.batch_size = batch_size
+        self.graph_client: GraphClient | None = None
+        self.requested_team_list: list[str] = teams
+
+    def load_credentials(self, credentials: dict[str, Any]) -> dict[str, Any] | None:
+        teams_client_id = credentials["teams_client_id"]
+        teams_client_secret = credentials["teams_client_secret"]
+        teams_directory_id = credentials["teams_directory_id"]
+
+        def _acquire_token_func() -> dict[str, Any]:
+            """
+            Acquire token via MSAL
+            """
+            authority_url = f"https://login.microsoftonline.com/{teams_directory_id}"
+            app = msal.ConfidentialClientApplication(
+                authority=authority_url,
+                client_id=teams_client_id,
+                client_credential=teams_client_secret,
+            )
+            token = app.acquire_token_for_client(
+                scopes=["https://graph.microsoft.com/.default"]
+            )
+            return token
+
+        self.graph_client = GraphClient(_acquire_token_func)
+        return None
+
+    def _get_threads_from_channel(
+        self,
+        channel: Channel,
+        start: datetime | None = None,
+        end: datetime | None = None,
+    ) -> list[list[ChatMessage]]:
+        # Ensure start and end are timezone-aware
+        if start and start.tzinfo is None:
+            start = start.replace(tzinfo=timezone.utc)
+        if end and end.tzinfo is None:
+            end = end.replace(tzinfo=timezone.utc)
+
+        query = channel.messages.get()
+        base_messages: list[ChatMessage] = query.execute_query()
+
+        threads: list[list[ChatMessage]] = []
+        for base_message in base_messages:
+            message_datetime = datetime.strptime(
+                base_message.properties["lastModifiedDateTime"], datetime_format_string
+            )
+
+            if start and message_datetime < start:
+                continue
+            if end and message_datetime > end:
+                continue
+
+            reply_query = base_message.replies.get_all()
+            replies = reply_query.execute_query()
+
+            # start a list containing the base message and its replies
+            thread: list[ChatMessage] = [base_message]
+            thread.extend(replies)
+
+            threads.append(thread)
+
+        return threads
+
+    def _get_channels_from_teams(
```

Comment:
Here and elsewhere, functions in the class that don't use self can either be pulled out, or if it makes the most sense to be semantically in the class class, it should be static.